### PR TITLE
Painless: Clean up PainlessMethod

### PR DESCRIPTION
**Pull request by jdconrad**
Renames and removes variables from PainlessMethod to follow the new naming convention.  Generates methodtypes at compile-time instead of using a method at run-time.  Moves write method to MethodWriter.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionRef.java [hunk 1]
```diff
@@ -138,37 +139,59 @@ public FunctionRef(Class<?> expected, PainlessMethod interfaceMethod, PainlessCo
...
+        } else if (delegateMethod.javaMethod.getDeclaringClass().isInterface()) {
             delegateInvokeType = H_INVOKEINTERFACE;
         } else {
             delegateInvokeType = H_INVOKEVIRTUAL;
         }
 
+        delegateMethodName = delegateMethod.javaMethod.getName();
+        this.delegateMethodType = delegateMethodType.dropParameterTypes(0, numCaptures);
+
+        this.interfaceMethod = interfaceMethod;
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionRef.java]: nit: can you use `this.` consistently when setting members in ctors here? It is difficult to tell what might be locals declared earlier in the method and what are members.

**Review by rjernst — APPROVED** [review]
LGTM

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionRef.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionRef.java]: Done.

**Comment by jdconrad** [comment]
@rjernst Thanks for the review!  Will commit as soon as CI passes.

**Pull request merged**